File: game/src/main/java/me/firecloud/game/rule/ActorNode.java
/**
 * 
 */
package me.firecloud.game.rule;

/**
 * @date 2012-10-12
 * 
 */
public abstract class ActorNode
    extends SingleNode
    implements Node {

  /*
   * (non-Javadoc)
   * 
   * @see me.firecloud.game.rule.Node#handle(me.firecloud.game.rule.Event)
   */
  public
      Node handle(Event event) throws RuleException {
    if (checkActor(event)) {
      // invoke next node
      return nextNode.handle(event);
    } else {
      return null;
    }
  }

  protected abstract
      boolean checkActor(Event event);

}

File: card/src/main/scala/me/firecloud/card/rule/ActorNode.scala
/**
 *
 */
package me.firecloud.card.rule

/**
 * @author kkppccdd
 * @email kkppccdd@gmail.com
 * @date Oct 19, 2012
 *
 */
abstract class ActorNode extends SingleNode {

  def handle(gameContext: GameContext, event: Event): Node = {
    if (checkActor(gameContext, event)) {
      return nextNode.handle(gameContext, event)
    } else {
      return null
    }
  }

  protected def checkActor(gameContext: GameContext, event: Event): Boolean
}